File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.py
def solution(friends, gifts):
    give = {i:[] for i in friends}
    receive = {i:[] for i in friends}
    gift_pont = {i:0 for i in friends}
    
    # gift_list = [[0] * len(friends)] * len(friends)
    gift_list = [[0] * len(friends) for i in range(len(friends))]
    
    friend_num = {}
    
    result = [0] * len(friends)
    
    for idx, friend in enumerate(friends):
        friend_num[friend] = idx
    
    for i in range(len(gifts)):
        giver, reciever = gifts[i].split(" ")
        
        gift_list[friend_num[giver]][friend_num[reciever]] += 1
        
        give[giver].append(reciever)
        receive[reciever].append(giver)
    
    
    for g, r, name in zip(give.values(), receive.values(), friends):
        gift_pont[name] = len(g) - len(r)
        
    
    for i in range(len(friends)):
        for j in range(len(friends)):
            if gift_list[i][j] > gift_list[j][i]:
                result[i] += 1
            
            elif gift_list[i][j] == gift_list[j][i]:
                if gift_pont[friends[i]] > gift_pont[friends[j]]:
                    result[i] += 1
    
    return max(result)
    

File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.java
import java.util.*;

class Solution {
    public int solution(String[] friends, String[] gifts) {
        Map<String, Integer> map = new HashMap<>();
        int answer = 0;
            
        for (int i = 0; i < friends.length; i++) {
            map.put(friends[i], i);
        }
        
        int[] index = new int[friends.length];
        int[][] record = new int[friends.length][friends.length];
        
        for (String s : gifts) {
            String[] tmp = s.split(" ");
            index[map.get(tmp[0])]++;
            index[map.get(tmp[1])]--;
            record[map.get(tmp[0])][map.get(tmp[1])]++;
        }
        
        for (int i = 0; i < friends.length; i++) {
            int cnt = 0;
            for(int j = 0; j < friends.length; j++) {
                if (i == j) continue;
                if (record[i][j] > record[j][i]) cnt ++;
                else if (record[i][j] == record[j][i] && index[i] > index[j]) cnt ++;
            }
            answer = Math.max(cnt, answer);
        }
        
        return answer;
    }
}